### PR TITLE
Examples and tests for module.

### PR DESCRIPTION
**Pull request by Artlvns**
chore: basic and full examples of usage.
chore: added extra validation of network_rules.ip_rules variable according to provider doc.
chore: integration and unit tests of module.

### CODE REVIEW

File: examples/02-full/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,20 @@
+output "container_id" {
+  description = "The ID of the storage container created for terraform backend state."
+  value       = module.state_container.container_id
+}
+
```
Comment (Laffs2k5) [inline examples/02-full/outputs.tf:5]:
```suggestion

output "container_name" {
  description = "Name of the storage container created for terraform backend state."
  value       = module.state_container.container_name
}

```

File: examples/01-basic/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,20 @@
+output "container_id" {
+  description = "The ID of the storage container created for terraform backend state."
+  value       = module.state_container.container_id
+}
+
```
Comment (Laffs2k5) [inline examples/01-basic/outputs.tf:5]:
```suggestion

output "container_name" {
  description = "Name of the storage container created for terraform backend state."
  value       = module.state_container.container_name
}

```

File: tests/integration-test-basic.tftest.hcl
Comment (Laffs2k5) [inline tests/integration-test-basic.tftest.hcl:1]: Consider using the naming module to make sure resource group names are unique across all integration tests.
Not important for "plan only", but then we are ready to also implement apply in the future ;-)

File: tests/integration-test-full.tftest.hcl
Comment (Laffs2k5) [inline tests/integration-test-full.tftest.hcl:1]: same comment for this test about using naming module

File: tests/unit-tests.tftest.hcl [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+        azure_region = ""
+    }
+
+    # Expect the plan to fail
+    expect_failures = [
+        var.azure_region,
+    ]
+}
+
+run "costcente_tag_value_can_not_be_empty" {
```
Comment (Laffs2k5) [inline tests/unit-tests.tftest.hcl]:
```suggestion
run "costcenter_tag_value_can_not_be_empty" {
```

File: tests/unit-tests.tftest.hcl [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+run "costcenter_tag_default_value" {
+    command = plan
+
+    assert {
+        error_message = "The 'costcenter_tag_value' default value is 142."
+        condition = azurerm_resource_group.tfstate.tags["costCenter"] == "142"
+    }
+}
+
+run "costcenter_tag_value_must_can_not_be_empty" {
```
Comment (Laffs2k5) [inline tests/unit-tests.tftest.hcl]:
```suggestion
run "costcenter_tag_value_can_not_be_empty" {
```

File: tests/unit-tests.tftest.hcl [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+        var.costcenter_tag_value,
+    ]
+}
+
+run "network_rules_default_action" {
+    command = plan
+
+    assert {
+        error_message = "The 'default_action' default value is Deny."
+        condition = azurerm_storage_account_network_rules.tfstate[0].default_action == "Deny"
```
Comment (Laffs2k5) [inline tests/unit-tests.tftest.hcl:171]: So nice to be able to verify this as part of the release 🤩

File: tests/unit-tests.tftest.hcl [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+
+run "network_rules_multiple_ip_rules" {
+    command = plan
+
+    variables {
+        network_rules = {
+            bypass                      = null
+            default_action              = "Deny"
+            ip_rules                    = ["224.0.0.1/28", "225.0.0.2"]
+            virtual_network_subnet_ids  = ["/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/Microsoft.Storage/storageAccounts/myStorageAccount"]
```
Comment (Laffs2k5) [inline tests/unit-tests.tftest.hcl]: Should this also be null?

```suggestion
            virtual_network_subnet_ids  = null
```

File: tests/unit-tests.tftest.hcl
Comment (Laffs2k5) [inline tests/unit-tests.tftest.hcl:1]: Wow! Nice job with writing unit tests 🏅

**Review by Laffs2k5 — APPROVED** [review]
Solid!

Now, how do we handle test and release for the [v0-branch](https://github.com/dsb-norge/terraform-azurerm-terraform-state-container/tree/v0-branch) ? Are we able to release from there and can maintain a variant of this test suite in that branch as well?

File: tests/unit-tests.tftest.hcl (hunk 6 above)
Comment (Artlvns) [reply tests/unit-tests.tftest.hcl]: basically it's up to personal preference. This particular test checks if we can add multiple ip addresses in ip_rules. What do you preffer ? or why do you want 'null' there ?

**Comment by Artlvns** [comment]
> Solid!
> 
> Now, how do we handle test and release for the [v0-branch](https://github.com/dsb-norge/terraform-azurerm-terraform-state-container/tree/v0-branch) ? Are we able to release from there and can maintain a variant of this test suite in that branch as well?

I think that these 3 points can help us to decide: 

- Do we plan support v0 in parallel ? 
- Do we need our TF module CI/CD to support parallel release from multiple branches, or it's rather the custom case ? 
- Are two points above worth the toil they demand, or we rather perform move off all state files to v1 in the long run?

File: tests/unit-tests.tftest.hcl (hunk 6 above)
Comment (Laffs2k5) [reply tests/unit-tests.tftest.hcl]: I just thought I saw a pattern where inputs was kept as minimal/undefined as possible and this one differed from the others. Not important, up to you ;-)

**Comment by Laffs2k5** [comment]
> I think that these 3 points can help us to decide:
> 
> * Do we plan support v0 in parallel ?
> * Do we need our TF module CI/CD to support parallel release from multiple branches, or it's rather the custom case ?
> * Are two points above worth the toil they demand, or we rather perform move off all state files to v1 in the long run?

I was hoping we could support parallel releases. Not a requirement as of now, further down the line though, I think we have to have a way. Imagine larger more complex modules introducing new major versions. We will be forced to support bug fixing and ensuring provider compatibility at least for a period after release.

As for this module and our current projects; We could keep the projects on the v0 branch using the git reference (as they are now, just point to this repo). Ideally we transition them over to v1, but I suspect that involves quite an effort 🤷

**Comment by Artlvns** [comment]
> > I think that these 3 points can help us to decide:
> > 
> > * Do we plan support v0 in parallel ?
> > * Do we need our TF module CI/CD to support parallel release from multiple branches, or it's rather the custom case ?
> > * Are two points above worth the toil they demand, or we rather perform move off all state files to v1 in the long run?
> 
> I was hoping we could support parallel releases. Not a requirement as of now, further down the line though, I think we have to have a way. Imagine larger more complex modules introducing new major versions. We will be forced to support bug fixing and ensuring provider compatibility at least for a period after release.
> 
> As for this module and our current projects; We could keep the projects on the v0 branch using the git reference (as they are now, just point to this repo). Ideally we transition them over to v1, but I suspect that involves quite an effort 🤷

Perhaps I understand release process wrongly then, but if we release V1 now with tag and GitHub release, and then v2 so you can still use V1. There is still problem with maintenance ofcourse, but by supporting several versions in a long run - we add more toil to us, so I thought that our goal will be always move to latest after new major release. 
Then we need to support branching in release action, I think there was config for that

**Pull request merged**